incubator/validator-core/src/main/java/io/aklivity/zilla/runtime/validator/core/StringEncoding.java

```diff
@@ -26,41 +26,24 @@ public boolean validate(
             int index,
             int length)
         {
-            int bytesIndex = 0;
-            while (bytesIndex < length)
+            final int limit = index + length;
+            while (index < limit)
             {
-                int numBytes;
-                if ((data.getByte(bytesIndex) & 0b10000000) == 0b00000000)
-                {
-                    numBytes = 1;
-                }
-                else if ((data.getByte(bytesIndex) & 0b11100000) == 0b11000000)
-                {
-                    numBytes = 2;
-                }
-                else if ((data.getByte(bytesIndex) & 0b11110000) == 0b11100000)
-                {
-                    numBytes = 3;
-                }
-                else if ((data.getByte(bytesIndex) & 0b11111000) == 0b11110000)
-                {
-                    numBytes = 4;
-                }
-                else
-                {
-                    break;
-                }
+                final int charByte0 = data.getByte(index);
+                final int charByteCount = (charByte0 & 0b1000_0000) != 0
+                    ? Integer.numberOfLeadingZeros((~charByte0 & 0xff) << 24)
+                    : 1;
 
-                for (int j = 1; j < numBytes; j++)
+                for (int j = 1; j < charByteCount; j++)
                 {
-                    if (bytesIndex + j >= length || (data.getByte(bytesIndex + j) & 0b11000000) != 0b10000000)
+                    if (index + j >= limit || (data.getByte(index + j) & 0b11000000) != 0b10000000)
                     {
                         break;
                     }
                 }
-                bytesIndex += numBytes;
+                index += charByteCount;
```

Comment:
We are calculating `index + j` more than once here and also `j` is not descriptive.
```
final int charByteLimit =. index + charByteCount;
for (int charByteIndex = index + 1; charByteIndex < charByteLimit; charByteIndex++)
{
    if (charByteIndex >= limit || (data.getByte(charByteIndex) & 0b11000000) != 0b10000000)
    {
        break validate;
    }
}
```
I think the `break` above needs to break from the outer `while` loop, not just the `for` loop, so we'll need to add a `validate:` label to the `while` loop, agree?